### ISSUE TITLE
NullableToStringCall should not trigger on platform types (ie. String!)

### ISSUE THREAD

**Issue by spand** [detekt/detekt#3571]
I just ran the rule on our main and largest project. The rule triggered on many cases but only a few were real bugs. Most were harmless platform typed strings. Thinking forward a bit then the only way to resolve this is to type the platform types away. (ie. `val foo: String = platformFoo`). Being forced to do that is something I feel really is an orthogonal concern. ie.: "Disallow any interaction with platform types other than assignment to a kotlin type."

## Expected Behavior of the rule
One of the cases that failed but really shouldn't.
```
val path = URLEncoder().encode("urlPath")
"https://www.google.com/$path"
```

## Context
As I said above failing on this is a bit extreme. A separate rule regarding stricter usage of platform types seems like the way to handle that case. I am very new to Detekt so I apologize if this is consistent with the rest of the rules.

**Comment by cortinico** [comment]
We do have another rule that is preventing Platform types but only on public APIs:
https://github.com/detekt/detekt/blob/master/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt

I see your point, but the rule is just checking if you're calling `toString()` on a nullable type (and on platform types as well).

Therefore, this is a valid case:
```
val path = URLEncoder().encode("urlPath")
"https://www.google.com/$path"
```

If you specify the type explicitly, Detekt won't raise an issue anymore but your code will fail if the platform call returns `null` so the code smell still stays.

I think you should adapt your code to encapsulate platform calls so that you return either a `String` or a `String?`.

**Comment by spand** [comment]
I get why you might want the strictness of this rule. What I experience is just that it is not very practical and it forces me to alter perfectly fine code to satisfy the rule. In my example `path` will never be `null` and there are many other methods in the JDK where the same is true.

What I am trying to say is that it is not really in the Kotlin spirit to treat platform types as nullable. The Kotlin way is to assume not nullable for external types. The Kotlin designers recognized in 2014 when they introduced platform types.

**Comment by BraisGabin** [comment]
I must say that I do prefer this code:

```kotlin
val path = URLEncoder().encode("urlPath")!! // or val path: String = URLEncoder().encode("urlPath")
"https://www.google.com/$path"
```

If you know that it can't be null the usage of `!!` is safe. And, if something happends, you get the error where expected.

Right know intellij has some warnings pointing out that you should not use platform types and that you should decide if you want to treat it as nullable or no nullable.

**Comment by cortinico** [comment]
> What I am trying to say is that it is not really in the Kotlin spirit to treat platform types as nullable.

I see your point. Though we use the compiler type resolution, and a platform type is marked as a nullable type. So in this case, treating platform type would mean to add an exception that goes against the compiler semantic.

@BraisGabin snippet is great for your use case.
Plus as I mentioned before, I would encapsulate those APIs into a Kotlin-friendly API. You have to handle platform types only in one place and you don't pollute your codebase with `@Suppress` annotations.

**Comment by spand** [comment]
I dont think I have any further arguments that might persuade you to do otherwise. It seems clear we diverge on what to expect from the rules but to sum up for any latecomers:

I would prefer if the rule was relaxed as described provided that a new rule would be created for being strict about platform types in the general case. The benefits of this is:
1) More in the spirit of Kotlin
2) Much more practical for introducing Detekt in an existing codebase
3) Existing behavior can still be achieved by using the new rule. A general strict rule about platform types also seem more in line with your arguments (though I guess this could be quite expensive to check).

Thanks for taking the time to discuss this 👍

**Comment by BraisGabin** [comment]
> Existing behavior can still be achieved by using the new rule. A general strict rule about platform types also seem more in line with your arguments (though I guess this could be quite expensive to check).

That's a really good point! You persuade me. We need a rule that ensure that you don't use platform types. And, if we have that rule, this rule should not check that, otherwise we could be raising two issues for the same problem.

**Comment by BraisGabin** [comment]
This change is already implemented in 1.19.0-RC1 so we can close this.
